File: solutions/erlang/rational-numbers/1/src/rational_numbers.erl
-module(rational_numbers).

-export([absolute/1, add/2, divide/2, exp/2, mul/2, reduce/1, sub/2]).


absolute({A, B}) -> reduce({abs(A), abs(B)}).

add({A1, B1}, {A2, B2}) -> reduce({A1 * B2 + A2 * B1, B1 * B2}).
sub({A1, B1}, {A2, B2}) -> reduce({A1 * B2 - A2 * B1, B1 * B2}).

mul({A1, B1}, {A2, B2}) -> reduce({A1 * A2, B1 * B2}).
divide({A1, B1}, {A2, B2}) -> reduce({A1 * B2, A2 * B1}).

exp({A, B}, N) when is_integer(N), N >= 0 -> reduce({pow(A, N), pow(B, N)});
exp({A, B}, N) when is_integer(N) -> reduce(exp({B, A}, abs(N)));
exp({A, B}, X) when is_float(X) -> reduce({math:pow(A, X), math:pow(B, X)});
exp(X, {A, B}) when is_float(X); is_integer(X) -> math:pow(math:pow(X, A), 1 / B).

reduce({A, B}) when B < 0 -> reduce({A * -1, B * -1});
reduce({A, B}) when is_integer(A), is_integer(B) -> 
  GCD = gcd(abs(A), abs(B)),
  {A div GCD, B div GCD};
reduce({A, _B}) when is_float(A) -> undefined;
reduce(R) -> R.

gcd(A, B) when A == 0; A == B -> B;
gcd(A, B) when A > B -> gcd(A - B, B);
gcd(A, B) -> gcd(B, A).

pow(X, N) when is_integer(N), N >= 0 -> pow(X, N, 1).

pow(_X, 0, Acc) -> Acc;
pow(X, N, Acc) -> pow(X, N - 1, X * Acc).

File: solutions/elixir/rational-numbers/1/lib/rational_numbers.ex
defmodule RationalNumbers do
  @type rational :: {integer, integer}

  @doc """
  Add two rational numbers
  """
  @spec add(r1 :: rational, r2 :: rational) :: rational
  def add(r1, r2) do
    {a1, b1} = r1
    {a2, b2} = r2
    reduce({a1*b2 + a2*b1, b1*b2})
  end

  @doc """
  Subtract two rational numbers
  """
  @spec subtract(r1 :: rational, r2 :: rational) :: rational
  def subtract({a1, b1}, {a2, b2}) do
    reduce({a1*b2 - a2*b1, b1*b2})
  end

  @doc """
  Multiply two rational numbers
  """
  @spec multiply(r1 :: rational, r2 :: rational) :: rational
  def multiply({a1, b1}, {a2, b2}) do
    {a, b} = reduce({a1*a2, b1*b2})
  end

  @doc """
  Divide two rational numbers
  """
  @spec divide_by(num :: rational, den :: rational) :: rational
  def divide_by({a1, b1}, {a2, b2}) do
    reduce({a1*b2, a2*b1})
  end

  @doc """
  Absolute value of a rational number
  """
  @spec abs(a :: rational) :: rational
  def abs({a, b}) do
    reduce({Kernel.abs(a), Kernel.abs(b)})
  end

  @doc """
  Exponentiation of a rational number by an integer
  """
  @spec pow_rational(a :: rational, n :: integer) :: rational
  def pow_rational({a, b}, n) when n > 0 do
    reduce({Integer.pow(a, n), Integer.pow(b,n)})
  end

  def pow_rational({a, b}, n) do
    m = Kernel.abs(n)
    reduce({Integer.pow(b, m), Integer.pow(a, m)})
  end

  @doc """
  Exponentiation of a real number by a rational number
  """
  @spec pow_real(x :: integer, n :: rational) :: float
  def pow_real(x, {a, b}) do
    :math.pow(x, a/b)
  end

  @doc """
  Reduce a rational number to its lowest terms
  """
  @spec reduce(r :: rational) :: rational
  def reduce(r) do
    {a, b} = r
    if a == 0 do
      {0, 1}
    else
      gcd = Integer.gcd(a, b)
      {r1, r2} = {div(a, gcd), div(b, gcd)}
      if r2 < 0 do
        {-r1, -r2}
      else
        {r1, r2}
      end
    end
  end
end
